### PR TITLE
Fix DSO memory layout

### PR DESCRIPTION
**Pull request by r1viollet**
# What does this PR do?

We were assuming that the elf layout was the same as the file layout.
This now registers the module in the correct address for the few cases where we do not match between file and elf layout.

Other minor adjustments:
- Find exe file using the proc exe link if we can't find it in the DSOs
- Adjust the symbol caching addresses

# Motivation

Fix infinite unwinding

# Additional Notes

Some amount of time was spent checking for impacts in whole host mode. Performances look the same (or slightly better)

# How to test the change?

Some unit tests are there, the rest of the tests should cover the unwinding.

### CODE REVIEW

File: include/ddprof_module.hpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  DDProfMod()
+      : _mod(nullptr), _low_addr(0), _high_addr(0),
+        _sym_bias(static_cast<Offset_t>(-1)), _status(kUnknown) {}
+
+  explicit DDProfMod(Status status) : DDProfMod() { _status = status; }
+
+  // In the current version of dwfl, the dwfl_module addresses are stable
+  Dwfl_Module *_mod;
+  ProcessAddress_t _low_addr;
```
Comment (r1viollet) [inline include/ddprof_module.hpp:34]: this should have been a range

File: include/ddprof_module.hpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
Comment (r1viollet) [inline include/ddprof_module.hpp:1]: we could unify naming of files (dwfl_modules /ddprof_module)

File: include/dso_symbol_lookup.hpp [hunk 3]
```diff
@@ -15,8 +15,9 @@ namespace ddprof {
 
 class DsoSymbolLookup {
 public:
-  SymbolIdx_t get_or_insert(ElfAddress_t addr, const Dso &dso,
-                            SymbolTable &symbol_table);
+  SymbolIdx_t get_or_insert(FileAddress_t normalized_addr, const Dso &dso,
```
Comment (r1viollet) [inline include/dso_symbol_lookup.hpp:18]: Feedback from Nicolas (not fully relevant to this PR): 
We should do the same as Go for the mappings 
- Single mapping per binary,
Low address is mod low (same for high).
Offset is 0

- We don't need to display these addresses (though they are nice to have within the pprof, for debug purposes)

**Comment by r1viollet** [comment]
General feedback : we need to add some test on following cases
- Position dependant
- Split debug (warning several variations)

File: src/dwfl_module.cc [hunk 4]
```diff
@@ -5,90 +5,76 @@
...
-        const char *mod_name = get_filename(main_name);
-        if (strcmp(mod_name, dso_name) != 0) {
-          // A dso replaced the mod - flag as incoherent for future release
-          LG_NTC("Incoherent DSO (%s) %lx != %lx dwfl_module (%s)",
-                 filepath.c_str(), dso._start - dso._pgoff,
-                 ddprof_mod._low_addr, main_name);
-          return DDProfMod(DDProfMod::kInconsistent);
-        }
-      }
+    if (ddprof_mod._low_addr != mod_range._low_addr) {
```
Comment (r1viollet) [inline src/dwfl_module.cc:33]: Feedback from Nicolas: This is an issue and yet we won't notice this.

File: src/dwfl_module.cc [hunk 5]
```diff
@@ -5,90 +5,76 @@
...
+      LG_NTC("Incoherent Modules (%s-%s) %lx != %lx dwfl_module)", module_name,
```
Comment (r1viollet) [inline src/dwfl_module.cc:34]: Test that when a new dso is mapped (within an existing mod_range), we update the ddprof_module

File: src/dwfl_symbol.cc [hunk 6]
```diff
@@ -40,74 +52,39 @@ bool symbol_get_from_dwfl(Dwfl_Module *mod, ProcessAddress_t process_pc,
...
-  } else {
-    symbol._lineno = 0;
+  if (line) {
+    int linep;
+    const char *localsrcpath =
+        dwfl_lineinfo(line, &process_pc, static_cast<int *>(&linep), 0, 0, 0);
+    if (localsrcpath) {
+      symbol._srcpath = std::string(localsrcpath);
+      symbol._lineno = static_cast<uint32_t>(linep);
+    }
```
Comment (r1viollet) [inline src/dwfl_symbol.cc:72]: do we need to errno ?

File: src/dwfl_module.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (r1viollet) [reply src/dwfl_module.cc:33]: So this is not correctly handled and this is not totally obvious
There is a difference between the lifetime of the sample and the state of the proc maps.

I think the best way is to list the files that were unloaded during a DSO update. If any of the unloaded DSOs had files linked, we should check for any related modules. 

Sadly, if these were loaded in the dwfl library, we will have to destroy the dwfl object.

File: include/ddprof_module.hpp (hunk 1 above)
Comment (r1viollet) [reply include/ddprof_module.hpp:34]: It might be better in the current form

File: include/dso_symbol_lookup.hpp (hunk 3 above)
Comment (r1viollet) [reply include/dso_symbol_lookup.hpp:18]: Opening a ticket to store information that allows us to symbolize after capture of stack traces

File: src/dwfl_module.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (r1viollet) [reply src/dwfl_module.cc]: Opened a ticket to follow up on this

Comment (r1viollet) [reply src/dwfl_module.cc]: This is a case where we need better telemetry

File: src/dwfl_symbol.cc (hunk 6 above)
Comment (r1viollet) [reply src/dwfl_symbol.cc:72]: Yes we do, APIs actually never mention this.

File: src/dwfl_module.cc (hunk 5 above)
Comment (r1viollet) [reply src/dwfl_module.cc]: We don't, follow up ticket.

File: src/dso_hdr.cc [hunk 7]
```diff
@@ -125,6 +124,25 @@ DsoHdr::DsoHdr() {
   _file_info_vector.emplace_back(FileInfo(), 0, true);
 }
 
+namespace {
+bool string_readlink(const char *path, std::string &link_name) {
+  char buff[1024];
+  ssize_t len = ::readlink(path, buff, sizeof(buff) - 1);
```
Comment (nsavoire) [inline src/dso_hdr.cc:130]: [Minor] You could use [std::filesystem::read_link](https://en.cppreference.com/w/cpp/filesystem/read_symlink)

**Review by nsavoire — APPROVED** [review]
Thanks for fixing this issue !

File: src/dso_hdr.cc (hunk 7 above)
Comment (r1viollet) [reply src/dso_hdr.cc:130]: - [ ] improve readlink using cpp API

**Pull request merged**